src/datahike/query.cljc

```diff
@@ -517,47 +536,56 @@
+  {:pre [(map? attrs-to-exclude)]}
+  (let [tuples (:tuples rel)
+        attrs (:attrs rel)
+        common-gtrs (map #(getter-fn attrs %) common-attrs)
+        keep-attrs (into [] (remove attrs-to-exclude) (keys attrs))
+        keep-idxs (to-array (map attrs keep-attrs))
+        key-fn (tuple-key-fn common-gtrs)]
+    {:tuples tuples
+     :attrs attrs
+     :keep-attrs keep-attrs
+     :keep-idxs keep-idxs
+     :key-fn key-fn}))
+
 (defn hash-join [rel1 rel2]
   (let [tuples1      (:tuples rel1)
         tuples2      (:tuples rel2)
         attrs1       (:attrs rel1)
         attrs2       (:attrs rel2)
-        common-attrs (vec (intersect-keys (:attrs rel1) (:attrs rel2)))
-        common-gtrs1 (map #(getter-fn attrs1 %) common-attrs)
-        common-gtrs2 (map #(getter-fn attrs2 %) common-attrs)
-        keep-attrs1  (keys attrs1)
-        keep-attrs2  (vec (set/difference (set (keys attrs2)) (set (keys attrs1))))
-        keep-idxs1   (to-array (map attrs1 keep-attrs1))
-        keep-idxs2   (to-array (map attrs2 keep-attrs2))
-        key-fn1      (tuple-key-fn common-gtrs1)
-        key-fn2      (tuple-key-fn common-gtrs2)]
+        common-attrs (vec (intersect-keys attrs1 attrs2))
+        params1 (make-hash-join-tuple-params rel1 common-attrs {})
+        params2 (make-hash-join-tuple-params rel2 common-attrs attrs1)]
     (if (< (count tuples1) (count tuples2))
-      (let [hash       (hash-attrs key-fn1 tuples1)
-            new-tuples (->>
-                        (reduce (fn [acc tuple2]
-                                  (let [key (key-fn2 tuple2)]
-                                    (if-some [tuples1 (get hash key)]
-                                      (reduce (fn [acc tuple1]
-                                                (conj! acc (join-tuples tuple1 keep-idxs1 tuple2 keep-idxs2)))
-                                              acc tuples1)
-                                      acc)))
-                                (transient []) tuples2)
-                        (persistent!))]
-        (Relation. (zipmap (concat keep-attrs1 keep-attrs2) (range))
-                   new-tuples))
-      (let [hash       (hash-attrs key-fn2 tuples2)
-            new-tuples (->>
-                        (reduce (fn [acc tuple1]
-                                  (let [key (key-fn1 tuple1)]
-                                    (if-some [tuples2 (get hash key)]
-                                      (reduce (fn [acc tuple2]
-                                                (conj! acc (join-tuples tuple1 keep-idxs1 tuple2 keep-idxs2)))
-                                              acc tuples2)
-                                      acc)))
-                                (transient []) tuples1)
-                        (persistent!))]
-        (Relation. (zipmap (concat keep-attrs1 keep-attrs2) (range))
-                   new-tuples)))))
+      (hash-join-tuples params1 params2)
```

Comment:
**Note**: I though `hash-join` had too much repetition so I refactored it. If you think this change is too unrelated to the rest of the code, I am happy to remove it from the PR.